_posts/2023-03-27-group-equivariant-cnn-2/group-equivariant-cnn-2.Rmd

```diff
@@ -0,0 +1,374 @@
+
+Today, we code a simple GCNN from scratch. Code and presentation tightly follow a [notebook](https://github.com/phlippe/uvadlc_notebooks/blob/master/docs/tutorial_notebooks/DL2/Geometric_deep_learning/tutorial1_regular_group_convolutions.ipynb) provided as part of University of Amsterdam's 2022 [Deep Learning Course](https://uvadl2c.github.io/). They can't be thanked enough for making available such excellent learning materials.
+
+In what follows, my intent is to explain the general thinking, and how the resulting architecture is built up from smaller modules, assigned a clear purpose each. For that reason, I won't reproduce all the code here; instead, I'll make use of the package [`gcnn`](https://github.com/skeydan/gcnn)`.` Its methods are heavily annotated; so to see some details, don't hesitate to look at the code.
+
+As of today, `gcnn` implements one symmetry group: $C_4$, the one that has been serving as a running example throughout post one. It is straightforwardly extensible, though, making use of class hierarchies throughout.
+
+## Step 1: The symmetry group $C_4$
+
+In coding a GCNN, the first thing we need to provide is an implementation of the symmetry group we'd like to use. Here, it is $C_4$, the four-element group that rotates by 90 degrees.
+
+We can ask `gcnn` to create one for us, and inspect its elements.
+
+```{r}
+# remotes::install_github("skeydan/gcnn")
+library(gcnn)
+library(torch)
+
+C_4 <- CyclicGroup(order = 4)
+elems <- C_4$elements()
+elems
+```
+
+    torch_tensor
+     0.0000
+     1.5708
+     3.1416
+     4.7124
+    [ CPUFloatType{4} ]
+
+Elements are straightforwardly represented by their respective rotation angles: $0$, $\frac{\pi}{2}$, $\pi$, and $\frac{3 \pi}{2}$.
+
+Groups are aware of the identity, and know how to construct an element's inverse:
+
+```{r}
+C_4$identity
+
+g1 <- elems[2]
+C_4$inverse(g1)
+```
+
+    torch_tensor
+     0
+    [ CPUFloatType{1} ]
+
+    torch_tensor
+    4.71239
+    [ CPUFloatType{} ]
+
+Here, what we care about most is the group elements' *action*. Implementation-wise, we need to distinguish between them acting on each other, and their action on the vector space $\mathbb{R}^2$, where our input images live. The former part is the easy one: It may simply be implemented by adding angles. In fact, this is what `gcnn` does when we ask it to let `g1` act on `g2`:
+
+```{r}
+g2 <- elems[3]
+
+# in C_4$left_action_on_H(), H stands for the symmetry group
+C_4$left_action_on_H(torch_tensor(g1)$unsqueeze(1), torch_tensor(g2)$unsqueeze(1))
+```
+
+    torch_tensor
+     4.7124
+    [ CPUFloatType{1,1} ]
+
+What's with the `unsqueeze()`s? Since $C_4$'s ultimate *raison d'être* is to be part of a neural network, `left_action_on_H()` works with batches of elements, not scalar tensors.
```

Comment:
I think I'm missing the conclusion here. Does `unsqueeze()` turn batches of elements back into scalar tensors?

Reply:
exactly!!
You got it though you don't even know torch (or do you?)  :-)))

It's really just for the reader who attentively looks at the code, or runs it - it needed to be there I think, and since the point gets taken up again, I guess it's important enough to not go into a footnote